### ISSUE TITLE
cygwin 32 bit

### ISSUE THREAD

**Issue by szepeviktor** [traviscross/mtr#55]
Could it be that on Cygwin `mtr: unable to get raw sockets.` means mtr cannot run on Windows?

**Comment by adpoliak** [comment]
This is still occurring.

How I built:
1. I downloaded the `0.87` release and unzipped it within the Cygwin home directory
2. I ran `./bootstrap.sh`
3. I ran `./configure --disable-ipv6 --without-gtk`
4. I edited `mtr.c` and `net.c` to comment out all instances of SCTP, as Windows does not have native support for that and Cygwin does not ship a compatibility layer for it either.
5. I edited `mtr.c` to bypass the UID 0 check on interval.
6. I ran `make`.
7. I used Microsoft's Manifest Tool (`mt.exe`) to add a manifest to `mtr.exe` to require running as an Administrator.

How I'm running:
1. I restarted Cygwin with Administrative privileges.
2. I ran `mtr ${TARGET_IP}`.
    The `ncurses` interface initialized successfully, but no output of the tracing process is shown on the screen.
3. I ran `mtr -bewz -c 10 -i 0.01 -s 1400 ${TARGET_IP}`.
    `mtr` runs to completion, but no report is actually produced.

**Comment by adpoliak** [comment]
Issue persists in commit c0bb08bd634360d3ece43fb84488c47cf3e39243 (current HEAD of `master` branch).

Build instructions:

1. Download and extract archive of commit in question within Cygwin environment.
2. Run `./bootstrap.sh`.
3. Run `./configure --disable-ipv6 --disable-bash-completion --without-gtk`.
4. Edit `mtr.c` to comment out UID 0 check within interval parameter processing.
5. Add `#define _BSD_SOURCE 1` to generated `config.h` file so `u_char`, `u_int`, `u_long`, `u_short`, `res_send_qhook`, `res_send_rhook`, and `timersub` are available in order for compilation/linking to complete successfully.
6. Use `mt.exe` to add manifest to `mtr.exe` to require running with Administrative privileges.

Slightly redacted output of executing `./mtr.exe -bewz -c 2 -i 0.5 -s 1400 4.2.2.1` follows:
![mtr-head-wide_report](https://cloud.githubusercontent.com/assets/3969913/18493231/5da841ce-79dd-11e6-96f1-3f6092c08780.png)

**Comment by szepeviktor** [comment]
@rewolff Could we have *official* Windows build instructions?

**Comment by rewolff** [comment]
I do not have any Windows machines. I get a headache when I have to use them. So if you guys manage to find a way that works, fine. I can try to add a thing or two that mean that you don't have to edit the sources again, but I can't test what I do. So I have to be careful to prevent breaking Linux, BSD, Solaris and others, while trying to make windows/cygwin work.

**Comment by matt-kimball** [comment]
I took a serious look at what it would take to get mtr running on Windows over the last two days, and I things look pretty dire.  The real-deal breaker is that I can't manage to receive any ICMP TTL-expired message, with one weird exception, which I'll get to in a moment.  Here are some of my observations:

1.  You'd want to use Cygwin as a base.  mtr uses select() as its main event loop, waiting for both incoming network packets, keyboard input, and timeouts.  Replacing this would be a lot of work, so Cygwin is pretty much the only way to go.

2.  The current code doesn't properly set up raw socket transmission for Windows.  First, Windows requires the sending socket to be bound to the interface you would like to send packets from.  mtr would need to figure out which interface is the right one to route to the destination, but the code in net_set_interfaceaddress_udp() already does this for the UDP casse.  You'd just need to re-purpose this code to be used for ICMP, TCP, etc.  This is pretty easy, and after hacking things up a bit, I had mtr sending the correct raw packets without too much trouble.

3.  For receiving, the first wrinkle, though, is that to receive any packets from a raw socket at all, you need to use WSAIoctl to set the receiving socket's SIO_RCVALL parameter to true.  This doesn't sound too hard, but WSAIoctl only works on Windows SOCKET handles, and Cygwin's space of socket fd's is a completely different animal than a SOCKET handle.  After some digging through the Cygwin source code, I found that you can use Cygwin's 'get_osfhandle' to get the underlying Windows handle from a Cygwin fd object.  So, putting that together, you can do something like `WSAIoctl(get_osfhandle(recvsock4), SIO_RCVALL, &true_opt, sizeof(true_opt), NULL, 0, &rsize, NULL, NULL)` to enable receiving on a raw socket.  (Who says the Windows API's aren't a joy to work with?  I kid, I kid.)

4.  With the above changes, I can no receive ICMP ECHO replies, but not TTL expired messages.  I was running Wireshark during my testing, and I could see the TTL expired messages arriving at the Ethernet adapter just fine.  As I was running mtr, my DSL modem and hosts beyond were all doing what they should be and sending back TTL expirations, but the Windows TCP/IP stack was filtering these out for some reason.  I found the `netsh interface ipv4 show ipstats` command, and I noticed that for every TTL-expired message received by windows, the counter for "In Unknown Protocol" incremented by one.  Very suspicious.  However, I know that Windows knows about ICMP, because it was happy to pass along the ICMP echo replies!

Here's where things get weird.  If I run tracert.exe (Windows' built-in traceroute), I can actually get _those_ TTL expired messages.  Even mtr can receive tracert.exe's TTL expired messages, but not any TTL expired messages in response to packets actually generated by mtr.  I thought Windows might have some hackery to indentify "good" TTL expired packets, so I started massaging the ICMP ids, sequence numbers and payload sizes until I was getting response packets that were bit-for-bit identical to the ones generated by tracert.exe's probes, as evidenced by Wireshark.  Still no dice.

I'm grasping at straws here, and I hope someone who knows more about the Windows TCP/IP stack can tell me I'm wrong, but I am beginning to suspect that Windows is doing early discards of any unexpected ICMP traffic that isn't in response to a ping that it knows it sent.  How would it be able to distinguish?  Well, Windows has this ICMP.DLL library built-in which is used to send and receive pings through a fairly limited API.  It's not a raw socket interface -- you can't send arbitrary packets -- but you can send ICMP echos with a particular TTL, and listen for their replies.  The only explanation for the weird behavior I cam seeing that I can think of is that Windows is whitelisting the TTL expired replies only for packets which were sent through ICMP.DLL.

This rasies the question, could mtr use ICMP.DLL?  Yes, I think it could, but the API is significantly different from the Unix-like socket API mtr is currently using, and it is limited to only ICMP, so you'd have to give up support for UDP, TCP and SCTP mtr traces if you did that.

Another, more powerful option, would be to bypass the Windows TCP/IP stack for receiving raw packets and use WinPcap to capture ethernet frames directly, and decode those to find your TTL replies.  That's kind of an interesting idea, because since WinPcap operates at the device driver level and timestamps packets when they are received in kernel-land, you'd get more precise timings on round trip time than mtr is currently able to provide.  (I'm not sure that any timing jitter due to user-space scheduling matters much in practice on machines which aren't under heavy CPU load, but for serious work, kernel level timings might be nice.)

However, either of these options would require some significant changes to the mtr code.  Getting mtr to work on Windows is going to be a matter of quite a bit more than just finding the right build instructions.

**Comment by matt-kimball** [comment]
I've done some work on getting a Windows port of mtr going, using.  I've got a running version of mtr, though it only support IP version 4, and it only supports ICMP probes.  (Not UDP, TCP, SCTP.)  

I made some compiled x86_64 binaries available here:  [https://github.com/matt-kimball/mtr/releases](https://github.com/matt-kimball/mtr/releases)

**Comment by sergeevabc** [comment]
@matt-kimball, is there a way to make a static release (single ```.exe```)?

**Comment by matt-kimball** [comment]
No.  The current architecture requires two .exe's, and the curses library requires external files to function.

**Comment by sergeevabc** [comment]
@matt-kimball, perhaps it’s very time to update Windows binary?
You compiled 0.87, whereas 0.92 is the latest.

**Comment by matt-kimball** [comment]
Sure thing.  Here is a binary built with the latest code:  https://github.com/matt-kimball/mtr/releases/tag/0.92-60e5

**Comment by sergeevabc** [comment]
@matt-kimball, thank you, sir.

**Comment by blackjack4494** [comment]
The windows version works like a charm so far. I haven't looked really into it. For my needs I have to use a modified version of mtr.

Tho I trying to port this c program to python with multiplatform support.

**Comment by matt-kimball** [comment]
@blackjack4494 -- I've been working on a Python package for using the backend to mtr ("mtr-packet") to send network probes.  It is cross platform.  You could use it to implement something like mtr in Python.

https://github.com/matt-kimball/mtr-packet-python

Take a look at "examples/trace-concurrent.py".  It is like a simple version of mtr.

Will that help you?

**Comment by blackjack4494** [comment]
@matt-kimball yes, kinda like that. my first step was also to utilize mtr-packet. But I would like to port mtr-packet to python as well. 
Your project right now is basically a pythron wrapper.
